datadog/dogstatsd/metrics.py

```diff
@@ -0,0 +1,77 @@
+from threading import Lock
+
+
+class MetricAggregator(object):
+    def __init__(self, name, tags, rate):
+        self.name = name
+        self.tags = tags
+        self.rate = rate
+
+    def aggregate(self, value):
+        raise NotImplementedError("Subclasses should implement this method.")
+
+    def flush_unsafe(self):
+        raise NotImplementedError("Subclasses should implement this method.")
+
+
+class CountMetric(MetricAggregator):
+    def __init__(self, name, value, tags, rate):
+        super(CountMetric, self).__init__(name, tags, rate)
+        self.value = value
+
+    def aggregate(self, v):
+        self.value += v
+
+    def flush_unsafe(self):
+        return {
+            "metric_type": "count",
+            "name": self.name,
+            "tags": self.tags,
+            "rate": self.rate,
+            "ivalue": self.value,
+        }
+
+
+class GaugeMetric(MetricAggregator):
+    def __init__(self, name, value, tags, rate):
+        super(GaugeMetric, self).__init__(name, tags, rate)
+        self.value = value
+
+    def aggregate(self, v):
+        self.value = v
+
+    def flush_unsafe(self):
+        return {
+            "metric_type": "gauge",
+            "name": self.name,
+            "tags": self.tags,
+            "rate": self.rate,
+            "fvalue": self.value,
+        }
+
+
+class SetMetric(MetricAggregator):
+    def __init__(self, name, value, tags, rate):
+        super(SetMetric, self).__init__(name, tags, rate)
+        self.data = set()
+        self.data.add(value)
+        self.lock = Lock()
```

Comment:
Similarly - maybe we should wait to add synchronization primitives until we actually have multithreaded code running. We can keep this PR focused to just introducing the metric classes and their associated properties.

Reply:
the serialization function needs values like namespace and container_id which is set in the base.py class. I think it would be more convenient to leave it as a function in base.py or move it to aggregator.py? Not sure.

This is what aggregator.py looks like so far https://github.com/DataDog/datadogpy/pull/838

Maybe we can move the serialization function there since it should have access to the DogStatsd client and its properties? It seems more intuitive that aggregator.py would handle serialization of a metric.